prog/dftb+/lib_dftbplus/mainapi.F90

```diff
@@ -188,10 +191,433 @@ function nrOfAtoms()
   end function nrOfAtoms
 
 
+  !> Check that the order of speciesName remains constant
+  !> Keeping speciesNames constant avoids the need to reset:
+  !> atomEigVal, referenceN0, speciesMass and SK parameters 
+  function checkSpeciesNames(inputSpeciesName) result(tSpeciesNameChanged)
+    use dftbp_initprogram, only: speciesName
+    use dftbp_accuracy, only: mc
+
+    !> Labels of atomic species from external program
+    character(mc),  allocatable, intent(in) :: inputSpeciesName(:)
+    !> Has speciesName changed?
+    logical                                 :: tSpeciesNameChanged
+    
+    tSpeciesNameChanged = any(speciesName /= inputSpeciesName)
+ 
+  end function checkSpeciesNames
+
+  
+  !> When order of atoms changes, update arrays containing atom type indices,
+  !> and all subsequent dependencies. Updated data returned via use statements
+  subroutine updateDataDependentOnSpeciesOrdering(env, inputSpecies)
+    !TODO(Alex) These will go to the top of the module
+    use dftbp_environment, only: TEnvironment
+    use dftbp_accuracy,    only: mc
+    use dftbp_initprogram, only: species0, mass, orb, iEqOrbitals, nIneqOrb, nMixElements, &
+         denseDesc, HSqrCplx, SSqrCplx, eigVecsCplx, HSqrReal, SSqrReal, eigVecsReal, &
+         q0, qShell0, qInput, qOutput, qInpRed, qOutRed, qDiffRed, nEl, speciesName
+
+    !> dftb+ environment 
+    type(TEnvironment),   intent(in) :: env
+    !> types of the atoms (nAllAtom) 
+    integer, allocatable, intent(in) :: inputSpecies(:)
+    
+    !Number of atoms must be conserved    
+    @:ASSERT(size(inputSpecies) == nAtom) 
+    species0 = inputSpecies
+    mass =  updateAtomicMasses(species0)
+    orb%nOrbAtom = updateAtomicOrbitals(species0)
+    call updateEquivalencyRelations(species0, iEqOrbitals, nIneqOrb, nMixElements)
+    call updateBLACSDecomposition(env, denseDesc)
+    call reallocateHSArrays(env, denseDesc, HSqrCplx, SSqrCplx, eigVecsCplx, HSqrReal, SSqrReal, eigVecsReal)
```

Comment:
These quantities only require updating if number of orbitals (hence atoms) changes. No need to call in this version of the API